### ISSUE TITLE
Allow users to filter in table column headers by changing OuiBasicTable

### ISSUE THREAD

**Issue by KrooshalUX** [opensearch-project/oui#209]
### Is your feature request related to a problem?
The established pattern for filtering table data does not meet modern use cases of filtering table values in a clean and practical way. Currently,`OuiFilterGroup` is used next to a search bar (if present) above a table - for each filterable column a button is added to the `OuiFilterGroup`. This increases visual noise, and visual noise increases cognitive processing, increasing friction when interacting with an already data-dense experience. Example taken from [Reporting ](https://playground.opensearch.org/app/reports-dashboards#/). 

<img width="998" alt="Screen Shot 2022-10-18 at 4 48 29 PM" src="https://user-images.githubusercontent.com/105819041/196565708-892ab53b-6e2f-469f-9df5-56625ae79447.png">

While this use case is limited to two filterable columns, but imagine if all columns were filterable, and what that interface may appear as.

### What solution would you like?
Make filters available in table column headers within `OuiBasicTable`
![image](https://user-images.githubusercontent.com/105819041/196784676-72b08df0-4905-4857-a82b-abf6036fd570.png)

A user clicks the filter icon (`OuiIcon type="filter"`) in a given column, and can, `onHover`, include (`OuiIcon: type="check"`) or exclude (`OuiIcon: type="cross"`) selects one or more dynamically populated filters (Either utilizing `OuiFilterSelectItem` if it can be used independently of `OuiFilterGroup`, or a new composed component which combines `OuiPopOver` + `OuiSelectable` + `OuiSearchBar`). Selections are applied `onClick`.
![image](https://user-images.githubusercontent.com/105819041/196784714-dca70e7f-5c38-4b73-b2fb-a157b4663c9c.png)
![image](https://user-images.githubusercontent.com/105819041/196784744-52b65658-cc5a-490c-ae54-5439812fe647.png)

If needed, a loading indicator would appear while the filter is applied (this may be a different experience between `OuiBasicTable` and `OuiInMemoryTable`). 

Filter icon changes state to filled to communicate there is a filter applied. Selected values are pinned to top of` OuiSelectable` when `OuiPopOver` is opened to reduce friction in deselecting a selected value. Additionally, a column-specific Reset and global table Reset functions are available. Note: this might require an additional glyph be added to OUI Icons to represent the active filter (`OuiIcon type="filterFilled"`)

![image](https://user-images.githubusercontent.com/105819041/196784895-2f40b33f-9588-4a98-ba41-f889f81c51bd.png)

Additional open question: Would it be useful to users to see a values count?
![image](https://user-images.githubusercontent.com/105819041/196784989-5c8352e8-28f6-4d94-b955-dd4f34c9978a.png)

### What alternatives have you considered?
Modifying `OuiFilterGroup` and nesting filters by column. Concerned that this is too high friction for filtering a table.

**Comment by BSFishy** [comment]
@KrooshalUX Could you make a new consolidated issue with specific action items for implementation of this?

**Comment by KrooshalUX** [comment]
@BSFishy yes that would be the next step after finalizing the design and goodnight through feature brief process.

**Comment by KrooshalUX** [comment]
@BSFishy the next step would be for engineering to take a pass at a meta issue with the implementation details based on the UX delivery. This will help us find any potential gaps in the UX designs and I can provide additional information when needed. Since this is our first time iterating on a component in this way, I am looking to this as a learning opportunity for all of us. Feel free to ask more questions here or in the engineering meta issue.

**Comment by AbhishekReddy1127** [comment]
@BSFishy @KrooshalUX 
I went through the above conversations, but I'm left with few questions.

1) By adding the column filter option, it does impact the current sorting behavior of the column. Right now the sorting happens when we click on the table header cell including the name, white space and sort icon. By implementing the filter option/icon, do we need to restrict the sorting only on the sort icon?

2) When does the global filter show up? Does it have to be there always or it has to be conditional? The reason why I am asking is because it would occupy a significant amount of space on the screen.

3) For the reset button, instead of showing it disabled, does it makes intuitive to show that button only when the user applies filters?

4) In mobile view, where should the filter options show up?

**Comment by kgcreative** [comment]
1. Sorting should be restricted to the name, or sort icon. Filtering will be only filter icon. We can remove the header as a tap target.
2. what do you mean by the global filter? I think this replaces that with column filters. Global filters would be for the dashboard level, and not really part of data grid. We would have a reset filters button adjacent to search, but that's it. I imagine that if you were to add filters via the search, then those would apply to the corresponding columns?
3. I think we're still forming opinions on this. I do like showing "reset filters" only when filters are applied.
4. We might need to think about the mobile view further. When I see it in the documentation site, i think we need to think about whether we show is as a drop down similar to "sorting", or if we do it in-line in each of the cards. This might need a bit more investigation. cc: @KrooshalUX 
<img width="554" alt="Screen Shot 2023-04-20 at 1 52 57 PM" src="https://user-images.githubusercontent.com/3443142/233485257-5527d342-9dfc-4d0a-8f39-cb15e9ee6ed9.png">

**Comment by KrooshalUX** [comment]
These are all great questions, thanks for taking a look into the proposal to help us identify gaps. 

Building on / responding to @kgcreative response:

1. It shouldn't impact the sort, no - for example if column a was sorted a>z and then i filtered column b to only show x values, the sort in column a would still be true. 
1a. For the interaction, however, I am looking into a hybrid button approach for sort and filter, however I am hesitant to remove the one-click-to-sort function that is currently present, its pretty nice. 
2. Correct this meant to replace most, if not all, implementations of using `OuiFilterGroup` or the `OuiSearchBar` filters that re commonly added by builders above the table (which is the scale problem we are having with new projects that require more columns than the filter bar can really handle). There are some edge cases here recently brought up in AI/ML that we need to take a deeper look into, so I will need some time on that from a pattern perspective, but that does not impact the implementation of these in-column filters. 
3. Possibly - at the time of publishing this proposal, we were still discussing the disabled pattern button. Since a lot of builders want to use / are utilizing incremental search, I am hoping for a one-button solution to clear filters and searches. I think we could probably get away with utilizing the same pattern they have here for sorting, but I will need to take a deeper look. We haven't studied much on mobile usage yet. 

I'll need some additional time on the shifts needed for the proposal to be rounded out. I will follow up here when I have some more concrete thoughts around these items.

**Comment by anirudha** [comment]
Closing — this issue has been blocked for over a year with no progress on its blockers. If the blocking conditions are resolved in the future, this can be reopened or a new issue created with updated context.
